### ISSUE TITLE
Postgres listener timeout when idle for too long

### ISSUE THREAD

**Issue by slashdotdash** [commanded/eventstore#127]
The `Postgrex.Connection` process may [silently die after being idle for too long](https://github.com/elixir-ecto/postgrex/issues/386) (e.g. one day of inactivity) which causes subscriptions to stop receiving events. 

This issue should be resolved in the next release of Postgrex which includes an idle_timeout for Postgrex.Notification to ping the connection [#388](https://github.com/elixir-ecto/postgrex/pull/388).

**Comment by slashdotdash** [comment]
@bamorim is investigating whether the latest Postgex master branch, which includes [#388](https://github.com/elixir-ecto/postgrex/pull/388), resolves this issue.

**Comment by bamorim** [comment]
@renanlage is taking a look at this.
Yesterday we tried to simulate the connection closing by closing the erlang port with:

```
{_, port} = \
  GenServer.whereis(EventStore.Notifications.Listener.Postgrex)\
  |> :sys.get_state()\
  |> Map.get(:pid)\
  |> :sys.get_state()\
  |> Map.get(:mod_state)\
  |> Map.get(:protocol)\
  |> Map.get(:sock)

:erlang.port_close(port)
```

We could successfully solve it by using the ping, but the point is that if a new event is received by the EventStore while the connection is down (before the ping) it loses the event (it is okay because the event is received when one more event arrives).

However, I was thinking in implementing a solution in EventStore itself, without depending on the Postgrex:

## Proposed solution

When starting the Listener, we also start a "Heartbeat" process that sends a NOTIFY with some specific body every `X` seconds. Then we implement some "timeout" mechanism that if we don't receive any response from the LISTEN connection we restart everything. This might be a good time to restart the EventHandlers as well, so we fetch new events from the database without waiting for a new event to arrive.

What do you think @slashdotdash?

**Comment by bamorim** [comment]
The point of implementing the heartbeat on the EventStore is that we don't need to wait Postgrex to deploy the new version.

**Comment by slashdotdash** [comment]
We could copy the [`Postgrex.Notifications` module](https://github.com/elixir-ecto/postgrex/blob/master/lib/postgrex/notifications.ex) into the event store (e.g. as `EventStore.Notifications`) to get early access to the idle timeout feature.

**Comment by bamorim** [comment]
@slashdotdash I think that would be nice. I'll try this out in my fork.

**Comment by andersonmcook** [comment]
I think we're running into this same problem, but after just minutes.  We have some reliable subscribers (coincidentally the ones that have the lowest `stream_id`s), but we have some that we think fall victim to this.

We can kill a subscriber that's behind and it will nearly immediately catch back up, only to eventually fall out of sync again.

**Comment by slashdotdash** [comment]
@andersonmcook A single Postgres database connection is used to listen to events and publishes them to all subscribers. If you're seeing some subscriptions are receiving events while others lag behind then it's unlikely to be caused by this issue. Feel free to open a separate issue with specific details of what you are seeing.

**Comment by andersonmcook** [comment]
@slashdotdash Thanks! We'll write something up with more explanation.

**Comment by bamorim** [comment]
I added a PR, but there is still a problem. If the sock dies, although Postgrex will reconnect, we don't issue a `Subscription.disconnect` and `Subscription.reconnect` in order to start the `:catch_up`  state again.
What that causes is that events lost during the time the sock was down are lost until we receive another events on the stream we subscribed.

**Comment by joaothallis** [comment]
Is it the same problem: https://github.com/commanded/commanded/issues/191?

**Comment by slashdotdash** [comment]
Fixed by #130.
